flink-runtime/src/main/java/org/apache/flink/runtime/operators/shipping/OutputEmitter.java

```diff
@@ -135,7 +116,7 @@ public OutputEmitter(ShipStrategyType strategy, TypeComparator<T> comparator, Pa
 		case PARTITION_CUSTOM:
 			return customPartition(record.getInstance(), numberOfChannels);
 		case PARTITION_RANGE:
-			return rangePartition(record.getInstance(), numberOfChannels);
+			return rangePartition((Tuple2<Integer, ?>)record.getInstance(), numberOfChannels);
```

Comment:
I would not make the `OutputEmitter` type specific, i.e., expect a `Tuple2<Integer, ?>`. Instead I would use a Integer DataDistribution, and a `TypeComparator<Tuple2<Integer,?>` that compares on field `f0`.